### PR TITLE
Autocomplete and Hinter

### LINKED ISSUE #99: p5.js library Autocomplete and/or link to specific parts of reference
Students requested a link to the p5.js reference in the editor.   Ideally we would also eventually have an option where you could highlight a function name like `fill`, right-click, and have a "find in reference" option.

### PR DESCRIPTION
**Pull request by peilingjiang**
Resolves #99

js-hinter

I was looking for WIPs and issues regarding autocompletion and found #99. Although the issue has been idle for a while, I think this feature can greatly help beginners to start and help reduce typos and increase efficiency in general. It turns out that CodeMirror provides a great base for implementing it so I tried to use its hinter addon and build this PR.

In addition to autocompletion, the hinters also have links to the corresponding documentation page. Not exactly the same as requested in #99 but provides a similar reference capability.

Some more details of this implementation:
- The style has been configured to work in all three themes.
- **TBD** It is triggered every time an alphabet is typed, unlike the native CodeMirror hinter that will only be triggered by `Ctrl-Space`.
- **TBD** It only works in `css` and `javascript` modes, as CodeMirror's HTML hinter works weirdly.
- **TBD** For JavaScript, it can only hint and autocomplete p5 keywords, unlike fancy IDEs that can also complete JS and customized keywords.
- I added a lightweight new dependency, [fuse.js](https://fusejs.io/), to tolerate some misspellings, e.g., you can still get an option of `circle` even when you type `cic` as a start.
- It depends on a new utils file - p5-hinter.js - that can be updated by update-p5-hinter.js (just like update-syntax-highlighting.js). It fetches the same data but organizes it in a different way.

I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] is from a uniquely-named feature branch and has been rebased on top of the latest `develop` branch. (If I was asked to make more changes, I have made sure to rebase onto `develop` then too)
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`

### CODE REVIEW

**Comment by catarak** [comment]
!!!!!!!!!!!! thank you for working on this! there's actually a google summer of code project focused on autocomplete for the web editor, so I'm going to let them take the charge in reviewing this and making sure the UI/UX is good/is web accessible/keyboard shortcuts are good/etc.

**Comment by peilingjiang** [comment]
> !!!!!!!!!!!! thank you for working on this! there's actually a google summer of code project focused on autocomplete for the web editor, so I'm going to let them take the charge in reviewing this and making sure the UI/UX is good/is web accessible/keyboard shortcuts are good/etc.

Didn't know about the ongoing progress. If possible, I'm also happy to work with the corresponding developer(s) to push it to the shipping state. Please let me know!

**Comment by catarak** [comment]
Thank you for continuing to update this PR! Sorry that progress has been slow. It turns out that there's not a GSoC project focusing on this issue, so I'm happy to review it.

**Comment by catarak** [comment]
Create Release Environment

**Comment by catarak** [comment]
Deployed at https://app-ted8cea-p5-js-web-editor.releaseapp.io

**Comment by peilingjiang** [comment]
> Thank you for continuing to update this PR! Sorry that progress has been slow. It turns out that there's not a GSoC project focusing on this issue, so I'm happy to review it.

Thanks! Please let me know anything that I need to put more work on!

**Comment by catarak** [comment]
> * **TBD** It is triggered every time an alphabet is typed, unlike the native CodeMirror hinter that will only be triggered by `Ctrl-Space`.
I think this is the right approach as it functions like more "normal" IDE's.

> * **TBD** It only works in `css` and `javascript` modes, as CodeMirror's HTML hinter works weirdly.
In what way does it work weirdly?

> * **TBD** For JavaScript, it can only hint and autocomplete p5 keywords, unlike fancy IDEs that can also complete JS and customized keywords.
Do you know how much work it would be to complete JS keywords and user-defined functions/keywords? This could be a separate PR. 

> * I added a lightweight new dependency, [fuse.js](https://fusejs.io/), to tolerate some misspellings, e.g., you can still get an option of `circle` even when you type `cic` as a start.
This is great!

> * It depends on a new utils file - p5-hinter.js - that can be updated by update-p5-hinter.js (just like update-syntax-highlighting.js). It fetches the same data but organizes it in a different way.
This is totally cool. I think a separate PR could document these scripts 😄

File: client/modules/IDE/components/EditorShowHint.jsx
Comment (catarak) [inline client/modules/IDE/components/EditorShowHint.jsx]: This file doesn't use JSX syntax so I think its extension should be just .`js`. I think you could call this file `show-hint.js` so it's obvious the CodeMirror file it is modified from.

Also, I was comparing your modified version with the CodeMirror version (see [diff](https://www.diffchecker.com/yAdquIjS) and I noticed that you removed a lot of the `aria-` attributes. These are really important to have for screen-reader users, is it possible to add them back with your UI changes?

File: client/styles/components/_hints.scss
Comment (catarak) [inline client/styles/components/_hints.scss]: You don't need to add the prefixes, this is automatically added with webpack.

File: client/styles/components/_hints.scss
Comment (catarak) [inline client/styles/components/_hints.scss]: You don't need to repeat the styles for each theme. You can use a SCSS mixin called `getThemifyVariable`. See [`_editor.scss`](https://github.com/processing/p5.js-web-editor/blob/develop/client/styles/components/_editor.scss)for an example for how to use it.

File: client/styles/components/_hints.scss
Comment (catarak) [inline client/styles/components/_hints.scss]: Again, you shouldn't need the prefixed versions.

File: client/styles/components/_hints.scss
Comment (catarak) [inline client/styles/components/_hints.scss]: I notice in this file you're using `px` sometimes, `em` sometimes, and `rem` sometimes. I prefer to use `rem`, but because I usually think in pixels, I'll use this trick for SCSS: `#{50 / $base-font-size}rem` where 50 is the measurement in pixels.

**Review by catarak — DISMISSED** [review]

**Comment by catarak** [comment]
One UI feedback—I think that the font size of the text in the hinter should be the same size as whatever the text size is set to in a user's preferences. This may need to be done in JS rather than CSS.

**Comment by catarak** [comment]
Overall this works so well and I can't wait to refine it and merge it in!!

**Comment by Atcold** [comment]
Hi @catarak, when were you thinking to merge this?

**Comment by peilingjiang** [comment]
> Hi @catarak, when were you thinking to merge this?

So sorry for the delay in updating the PR! I don't think I'll have time to resolve the problems until early September as I'm running for some submission deadlines. I'll try to update as soon as possible for it to be ready to merge.

**Comment by catarak** [comment]
> So sorry for the delay in updating the PR! I don't think I'll have time to resolve the problems until early September as I'm running for some submission deadlines. I'll try to update as soon as possible for it to be ready to merge.

No worries at all! I really appreciate the work you have done so far ✨

File: client/modules/IDE/components/EditorShowHint.jsx
Comment (peilingjiang) [reply client/modules/IDE/components/EditorShowHint.jsx]: I don't remember removing them. Is it possible that you were comparing to a different version of their original file? You can see that all `aria` related lines miss the semicolon, which indicates they might add them later? But yeah, they are good, so I added them to the file. The file was also renamed.

File: client/styles/components/_hints.scss
Comment (peilingjiang) [reply client/styles/components/_hints.scss]: Resolved with other requested CSS changes!

**Comment by peilingjiang** [comment]
Hi @catarak Can you please review the latest updates? All the requested changes are made, except the final UI improvement of dynamic text size which I think would take some more time for me to finish. Would it be possible for us to merge the current one so that people can start using it and I can work on the dynamic size with another PR?

**Comment by catarak** [comment]
> Hi @catarak Can you please review the latest updates? All the requested changes are made, except the final UI improvement of dynamic text size which I think would take some more time for me to finish. Would it be possible for us to merge the current one so that people can start using it and I can work on the dynamic size with another PR?

This makes sense to me! I'll review it right now.

**Comment by catarak** [comment]
Just two things:
1. I noticed a couple of lines removed from the CodeMirror `show-hint.js` file to adapt it to work in the p5.js Editor. I'm not sure what they do, but I'm just curious if there's a reason you took them out.
2. I also think there should be a way to open the reference pages without using your mouse. I'm going to think about how this should work, because the usual way to focus on the button would be to use <kbd>Tab</kbd>. This obviously doesn't work because it triggers the autocomplete. If you have any ideas, let me know!

Edit: I've been thinking about (2) and I haven't been able to figure it out! I can't find any comparable UX patterns.

**Comment by catarak** [comment]
I couldn't figure out how to push to this branch so I made a couple of commits to https://github.com/processing/p5.js-web-editor/tree/catarak/autocomplete

**Comment by peilingjiang** [comment]
Thank you! I will rebase mine to yours. I think I removed the lines as I found the `this.id` is undefined, but I just realized that they do have a line to set that and I didn't add it at the beginning. Thanks for the updates! I am looking into the Tab functionality right now.

**Comment by peilingjiang** [comment]
> Just two things:
> 
> 1. I noticed a couple of lines removed from the CodeMirror `show-hint.js` file to adapt it to work in the p5.js Editor. I'm not sure what they do, but I'm just curious if there's a reason you took them out.
> 2. I also think there should be a way to open the reference pages without using your mouse. I'm going to think about how this should work, because the usual way to focus on the button would be to use Tab. This obviously doesn't work because it triggers the autocomplete. If you have any ideas, let me know!
> 
> Edit: I've been thinking about (2) and I haven't been able to figure it out! I can't find any comparable UX patterns.

I added the Tab. The thing is once you focus on the link, the hinter is blurred so up-n-down arrows no longer work - if you then change your mind and don't want to go to the documentation page and want to switch to another item, you'll need to click. Alternatively, as the second last commit did, I can also trigger the link directly by pressing Tab, we may want to make it as Ctrl/Command + Tab though (which is possible). I think I'd prefer this way. Please let me know what do you think!

**Comment by catarak** [comment]
Thanks for making these changes! I think this is moving in the right direction but it's not quite there.

I'm not sure about using <kbd>Tab</kbd> to open the reference link. Some ideas:
- Use the right arrow key to focus on the arrow, then <kbd>Enter</kbd> to open the link. Then left arrow would allow you to move up and down in the autocomplete again.
- Add a new keyboard shortcut (perhaps <kbd>Shift</kbd> + Right Arrow) that opens the selected dropdown item reference in a new tab, which keeps your focus on the editor window. This would allow a user to open multiple reference tabs, which could be useful. We could document this in the keyboard shortcuts. 

Thank you so much for all of your work on this! Can't wait to get this right and merge it in ✨

**Comment by peilingjiang** [comment]
Great suggestions. I added arrow navigations - Right Arrow brings the nav to links, and Left Arrow brings it back. When a link is highlighted, hitting Return will not complete the word, but go to the page instead.

However, for Shift-Right Arrow shortcut, I did research and tried multiple ways, but it seems browsers intentionally disallow you to use JS to open a new tab in the background, to avoid malicious websites opening something without the user knowing it. I still added it, but the new tab will always be focused instead of stacking in the background. When the new tab is closed, the hinter is still focused.

On the other hand, Ctrl/Command + click still works for opening tabs in the background.

**Comment by catarak** [comment]
Thank you for making that change, I think the arrow keys work great!

One last thing I'm noticing is the color contrast. For the light and dark themes, I've tried to make everything WCAG 2.0 compliant, and the high contrast theme WCAG 3.0 compliant. I usually use this [contrast checker](https://webaim.org/resources/contrastchecker/). I'm happy to help with this, just let me know!

**Comment by peilingjiang** [comment]
> Thank you for making that change, I think the arrow keys work great!
> 
> One last thing I'm noticing is the color contrast. For the light and dark themes, I've tried to make everything WCAG 2.0 compliant, and the high contrast theme WCAG 3.0 compliant. I usually use this [contrast checker](https://webaim.org/resources/contrastchecker/). I'm happy to help with this, just let me know!

I've made several changes to increase the readability and checked with WCAG, but this is relatively new to me so please feel free to tweak colors and UI if needed!

**Comment by Atcold** [comment]
I just wanted to point out that Jupyter lets you access the documentation with Shift-Tab.
Then, I didn't get the right / left arrow bit. Usually you go up / down, no?

**Comment by catarak** [comment]
> I just wanted to point out that Jupyter lets you access the documentation with Shift-Tab. Then, I didn't get the right / left arrow bit. Usually you go up / down, no?

Oh, thanks for the reference! I'll try Jupyter and see if I like how the autocomplete/opening documentation works.

**Comment by catarak** [comment]
@peilingjiang the color changes you made look great. the last things i would like to add to this are:
1. Adding the autocomplete keyboard usage to the keyboard shortcuts page
2. Adding the ability to turn on/off the the autocomplete in the preferences.

I can definitely work on these too!

**Comment by peilingjiang** [comment]
@catarak For these, do I also need to modify the translation files for each language?

**Comment by catarak** [comment]
> @catarak For these, do I also need to modify the translation files for each language?

No! I think those can be separate tasks.

**Comment by peilingjiang** [comment]
I added it to the shortcut page, but for the switch I'm not sure how to do it, can you help me with that? @catarak I can rebase my branch to yours once you are done. Thanks.

**Comment by peilingjiang** [comment]
Also not sure why the test failed, all I did was add a piece of static HTML.

**Comment by catarak** [comment]
> I added it to the shortcut page, but for the switch I'm not sure how to do it, can you help me with that? @catarak I can rebase my branch to yours once you are done. Thanks.

Yes absolutely!

> Also not sure why the test failed, all I did was add a piece of static HTML.
I think this is due to a snapshot test failing—I can fix this!

**Comment by Atcold** [comment]
@catarak, did you manage to fix the failing test?

**Comment by peilingjiang** [comment]
Hey @catarak are there recent updates on the failing test and turning on/off feature?

**Comment by Atcold** [comment]
@catarak, can you merge this PR please?

**Comment by peilingjiang** [comment]
Hey @catarak any updates here?

**Comment by Atcold** [comment]
We're close to one year anniversary for this PR.
Can we please merge it? Thanks.

**Comment by catarak** [comment]
Apologies for this taking so long to merge in! I clearly reached a place of burnout and am in the process of finding new leadership (see #2010). That being said, I am doing a lot better, and am focusing my attention on updating packages, documentation, closing pull requests, and organizing issues, so I will merge this one within the next month.

**Comment by Atcold** [comment]
Yay! One month anniversary today! 🥹
Jokes aside, I hope you're doing well 😊

**Comment by peilingjiang** [comment]
Any plans to reboot the merging of this PR? @catarak @raclim

**Comment by raclim** [comment]
Hi @peilingjiang! Thank you so much for your patience and understanding during our transitional period. I'm still learning the ropes for this position so merging in PRs and deployment might take some time, but I'd definitely love to continue the efforts that went into this feature. 

Timeline wise, I'm hoping to get a flow established over the following month and will be delving into the PRs (including this one) within that time. 🙂

**Comment by peilingjiang** [comment]
Hey @raclim any plan on merging this? Thanks!

**Comment by raclim** [comment]
Hi @peilingjiang! I'm really sorry for not getting to this 😅 I'm currently away until the end of the week so I won't be as active here until then, but I'd still love to merge your work in! Would you be able to resolve some of the conflicts in this branch?

**Comment by peilingjiang** [comment]
@raclim I have resolved the conflicts, added the switch to turn it on and off, added tests for the switch, and updated the UI a little. It should be all ready to merge now.

Screenshot 2023-05-25 at 15 36 27

**Comment by Atcold** [comment]
Let's do it! I'm waiting for this to be merged to start teaching P5.js using the native editor! 🥳🥳🥳

**Comment by peilingjiang** [comment]
@raclim Please let me know if there's anything that needs an update! Thanks for the review!

**Comment by raclim** [comment]
@peilingjiang Thanks so much for continuing to work on this! I just got back so I'll start looking over this today, and will hopefully get back to you on this by tomorrow!

File: client/modules/IDE/components/show-hint.js
Comment (raclim) [inline client/modules/IDE/components/show-hint.js:208]: Let me know if this might've been an issue on my end, but this feature wouldn't work for me until `completion.item.text` was updated to `completion.text`!

File: client/modules/IDE/components/Editor.jsx
Comment (raclim) [inline client/modules/IDE/components/Editor.jsx]: It would be great to remove some unused, commented out lines like these!

**Review by raclim — COMMENTED** [review]
Hi! Overall, I feel like this looks really great so far! :) 

As noted in my other comments, I had some difficulty activating the feature without the noted line change within the `show-hint.js` file. Visually, it also appeared a bit differently for me. 

Screenshot 2023-06-01 at 5 13 54 PM

Please let me know if there was something I might've missed or done incorrectly while reviewing this, or go ahead with adding any updates, thanks!

File: client/modules/IDE/components/Editor.jsx
Comment (peilingjiang) [reply client/modules/IDE/components/Editor.jsx]: Removed!

File: client/modules/IDE/components/show-hint.js
Comment (peilingjiang) [reply client/modules/IDE/components/show-hint.js:208]: Hi @raclim thanks for trying it out. I couldn't replicate or locate the issue. The only difference that I can think of between our environments is the `node_modules` folder, but you can also see that the only changes in the `package-lock.json` are all about the `fuse.js` package. Nevertheless, if you could try a clean reinstall of the modules, it might help.

The interface looks like the version for CSS (activated for `.css` files) - no links to the references, just completions. But it shouldn't be the case when a JS file is activated. I tried to change some code to ensure the correct mode is used, but don't know if it works as I cannot replicate the problem locally.

Would it be possible for you to create a release environment like [this](https://github.com/processing/p5.js-web-editor/pull/1873#issuecomment-881003367)? That would also be very helpful!

Comment (raclim) [reply client/modules/IDE/components/show-hint.js:208]: Thanks for those suggestions, you were right I was behind a few packages 😂 Works great for me now!

**Review by raclim — APPROVED** [review]

**Comment by raclim** [comment]
I looked through it with the updates and think it looks great! I also created a [release environment](https://app-ted3a10.release.editor.p5js.org/) for it. Thanks again so much for your hard work and super excited to see it live soon! :)

**Comment by Atcold** [comment]
Yay! 🥳🥳🥳

@peilingjiang, is there a way to trigger a command syntax suggestion?
Say I type `re`, the suggestion `rect` is provided, but there is no mention of what arguments it expects.
Having something like `rect(x, y, w, [h], [tl], [tr], [br], [bl])` would help a lot!

**Comment by peilingjiang** [comment]
Thanks a lot @raclim @catarak @Atcold for all the support and comments! Pretty excited to see it being merged!

@Atcold Great suggestion! And here you go #2236 : )

**Pull request merged**
